docs/thinking_in_jax.md

```diff
@@ -0,0 +1,401 @@
+---
+jupytext:
+  text_representation:
+    format_name: myst
+kernelspec:
+  display_name: Python 3
+  name: python3
+---
+
+```{code-cell}
+:tags: [hide-cell]
+# Configure ipython to hide long tracebacks.
+import sys
+ipython = get_ipython()
+
+def hide_traceback(exc_tuple=None, filename=None, tb_offset=None,
+                   exception_only=False, running_compiled_code=False):
+    etype, value, tb = sys.exc_info()
+    value.__cause__ = None  # suppress chained exceptions
+    return ipython._showtraceback(etype, value, ipython.InteractiveTB.get_exception_only(etype, value))
+
+ipython.showtraceback = hide_traceback
+```
+
+
+# How to think in JAX
+
+JAX provides a simple and powerful API for writing accelerated numerical code,
+but working effectively in JAX sometimes requires extra consideration.
+This document is meant to help build a ground-up understanding of how JAX operates,
+so that you can use it more effectively.
+
+## JAX vs. NumPy
+
+**Key Concepts:**
+
+- JAX provides a NumPy-inspired interface for convenience.
+- Through duck-typing, JAX arrays can often be used as drop-in replacements of NumPy arrays.
+- Unlike NumPy arrays, JAX arrays are always immutable.
+
+NumPy provides a well-known, powerful API for working with numerical data.
+For convenience, JAX includes the {mod}`jax.numpy` module, which closely mirrors the
+NumPy API and provides easy entry into JAX. Here are some operations you might do in ``numpy``:
+
+```{code-cell}
+import matplotlib.pyplot as plt
+
+import numpy as np
+x = np.linspace(0, 10, 1000)
+y = 2 * np.sin(x) * np.cos(x)
+plt.plot(x, y);
+```
+
+And here is the equivalent with {mod}`jax.numpy`:
+
+```{code-cell}
+import jax.numpy as jnp
+x = jnp.linspace(0, 10, 1000)
+y = 2 * jnp.sin(x) * jnp.cos(x)
+plt.plot(x, y);
+```
+
+The code blocks are identical aside from replacing ``np`` with ``jnp``, and the results are the same.
+As we can see, JAX arrays can often be used directly in place of NumPy arrays for things like plotting.
+
+The arrays themselves are implemented as different Python types:
+
+```{code-cell}
+type(np.arange(10))
+```
+
+```{code-cell}
+type(jnp.arange(10))
+```
+
+Python's [duck-typing](https://en.wikipedia.org/wiki/Duck_typing) allows JAX arrays and NumPy arrays
+to be used interchangeably in many places.
+
+However, there is one important difference between JAX and NumPy arrays: JAX arrays are immutable,
+meaning that once created their contents cannot be changed.
+
+Here is an example of mutating an array in Numpy:
+
+```{code-cell}
+x = np.arange(10)  # Numpy: mutable arrays
```

Comment:
Sorry, I missed these two during the previous check:

```suggestion
Here is an example of mutating an array in NumPy:

```{code-cell}
x = np.arange(10)  # NumPy: mutable arrays
```

Reply:
Done, thanks